### ISSUE TITLE
Refactor handler(Raw|Http) calls to use Promises

### ISSUE THREAD

**Issue by pocesar** [pocesar/node-jsonrpc2#29]
To the implementators and extensions POV, refactoring a standard way using promises in those methods increase testability, interop and easier integrations. It also decouples the functionality that is currently ingrained inside those methods, and expose them for better unit testing

**Comment by danielmihai** [comment]
We could split this part into two, to get some speed for reaching the milestone earlier.

**Comment by pocesar** [comment]
I'll push some of the things I've done so far, what is nagging me is the parser inside the handler methods

**Comment by danielmihai** [comment]
In what way ? what's the bugging part?

**Comment by danielmihai** [comment]
i synced the fork with this upstream, but I see no changes or commits.

**Comment by pocesar** [comment]
I didn't push yet, it's a mess (only 4 tests passing)

**Comment by pocesar** [comment]
ok I put it in the develop branch, it's really broken right now. some places are creating internal classes inside, and expect them to `handleMessage`. I can either pass the conn to the promise result (would be better for control, and for integration) or I can just handle everything the way it is right now, and *just* resolve the promise when it's done

**Comment by danielmihai** [comment]
I think I have some free time this weekend, after this I'm going to start working again. If you think you need a hand, just say, and I'll jump in.

Cheers! Onto the v2.0 of json-rpc2!

**Comment by pocesar** [comment]
alright, if you can, I won't be able to work on it from today through sunday though. one thing, I removed the Promise.method from inside the checkAuth because it's an expensive call (it uses new Function under the hood), so having it in a frequently called function could have performance implications

**Comment by danielmihai** [comment]
What if we use it only once, when enabling the authorization handler type
(enableBasic/Cookie/JWTAuth) instead of every time we check the
authorization headers? I think this might solve the performance issue and
still be able to support synchronous implementations of the handler?

**Comment by danielmihai** [comment]
Would you mind if I try a clean implementation (from where I left off) ?

If not, I will try to continue from the last commit.

**Comment by pocesar** [comment]
yeah, that's what I did, moved the Promise.method call inside the `enable*` functions (which could in turn become a private method that only does that `this.authHandler = Promise.method(handler)`

**Comment by danielmihai** [comment]
Okay, I was mistaken. If we set it inside the `enable*` functions once, I think it's enough. If we set it once during the enable functions, the `setAuthType` will do another unnecessary wrap each time we switch. I think we can remove that part. What is your opinion ?

**Comment by danielmihai** [comment]
any progress on this issue ?

**Comment by pocesar** [comment]
nope, I'm having to finish off a platform for a startup with a tight deadline. will resume it hopefully by this weekend, but if you want to work on it, go ahead

**Comment by danielmihai** [comment]
I managed to get a working form for handleHttp and get the jsonrpc-test.js tests working (refactored them the 'promise way' i guess). 
I might be able to make the other tests run until tomorrow evening (Bucharest time), too tired now.

**Comment by danielmihai** [comment]
I managed to make the authorization tests run too. We have to make a discussion in the case the server and client have different auth types, what kind of error do we return (InternalError or InvalidParams with the error message provided in the auth handler). I checked if the message was UNAUTHORIZED (const) return InvalidParams, else return InternalError with the message provided by the developer in the handler. 
(line 350 in the new pull request).

I will submit an early pull request to continue the discussion there for refactoring/improvements etc.

Thanks!

**Comment by danielmihai** [comment]
Are you going to take over the progress on the issue, including merge change requests and modifications? I've been uber busy these days, uni, work and bucharestjs hacks (on open data, first prize by the way ;) ) Let's have another talk on thursday (it's the first day I would be free), what do you say ?

**Comment by pocesar** [comment]
I merged your PR so I would be able to continue on it (unless you made some new changes). I'm almost done with a sprint here for a private project and will resume working it in the meantime. Congrats on first prize! 
Depending on the time the you're available, we can match the timezones (GMT-2 day time right now)

**Comment by danielmihai** [comment]
Hi @pocesar ,

Do you want to set up a time interval to have the talk ? If you are still up for it and have the time today. 
I don't know how to reach you in a more immediate manner except github mentions. Gitter sends email notifications like github, so there is no difference, unless you install the mobile app. 

Awaiting for your answer in, let's say, 2 hours (maximum 3)? It'll be the latest 10pm for me. Otherwise, schedule it for tomorrow and, if you have any changes, could you please push them on the remote ?

All the best,
Daniel.
